Reintroduce poisoning on transaction restarts

tbg opened the pull request:
When pushing an intent timestamp forward, we now unconditionally poison
the sequence cache. When checking the sequence cache, a SNAPSHOT txn
ignores the poisoned entry.

We chose this route since we don't know at time of the poisoning (which
happens as a side-effect of `ResolveIntent{,Range}` whether the other
transaction is `SNAPSHOT`. An alternative to this change is adding the
ISOLATION field to `TxnMeta`; in light of the upcoming proto freeze
we should consider whether there are other reasons for adding this field
to TxnMeta (in which case this change could be simplified).

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/4526)

<!-- Reviewable:end -->

tamird commented:
Reviewed 8 of 8 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[storage/replica.go, line 1432 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4526#-KB5V5XHfwug4W9mdIlX:-KB5V5XHfwug4W9mdIlY:-870186560) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b7676b33cc72704099363115a768f030b89ffde/storage/replica.go#L1432)):*
any reason not to call `readErr` `err`?

---

*[storage/replica_test.go, line 2138 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4526#-KB5XDfO5KZ4f_PZGRdh:-KB5XDfO5KZ4f_PZGRdi:-365430495) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b7676b33cc72704099363115a768f030b89ffde/storage/replica_test.go#L2138)):*
s/They/they/

---

*[storage/sequence_cache.go, line 113 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4526#-KB5WtqruJr_oxpuHJjP:-KB5WtqsM84HUaXT6z5i:1634761571) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b7676b33cc72704099363115a768f030b89ffde/storage/sequence_cache.go#L113)):*
isn't this a better fit for `MVCCIterate` now?

---

*[storage/sequence_cache_test.go, line 146 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4526#-KB5X6U70rcau1e8voBr:-KB5X6U70rcau1e8voBs:251836990) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b7676b33cc72704099363115a768f030b89ffde/storage/sequence_cache_test.go#L146)):*
i think it's time to sprinkle some comments in this test

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4526)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
LGTM, but I think I'd prefer the alternate solution in which we store the transaction isolation level in the intent. If we have to make this kind of distinction then it feels better to me to have the poisoner know what kind of transaction it's dealing with than to do the "wrong" thing and have the other transaction ignore it later. Is there any reason not to store the isolation level here than the small space costs?

---

Review status: 3 of 8 files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4526#-:-KB5bNnxjnjEzcdAPgVd:1968221126)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 5 of 5 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4526)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
@bdarnell not that I know of. @petermattis? I also think simple goes a long way. Don't think this turned out too bad, but it's a wrinkle of complexity that could simply be avoided.

---

Review status: 5 of 8 files reviewed at latest revision, 4 unresolved discussions.

---

*[storage/replica.go, line 1432 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4526#-KB5V5XHfwug4W9mdIlX:-KB5bVVxk2ivo48Ynt21:-1389969120) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b7676b33cc72704099363115a768f030b89ffde/storage/replica.go#L1432)):*
there's a local `err` just below. Seems kind of funky reusing it.

---

*[storage/sequence_cache.go, line 113 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4526#-KB5WtqruJr_oxpuHJjP:-KB5fDjCc2_FZEQAWNMF:-1238309134) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b7676b33cc72704099363115a768f030b89ffde/storage/sequence_cache.go#L113)):*
Probably. Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4526#-:-KB5gQCQZYOmNGj-iqZ7:1587177867)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[storage/sequence_cache.go, line 110 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4526#-KB5hEmcyrHN-fh5F-Y5:-KB5hEmd97G_SOS24neS:704684290) ([raw file](https://github.com/cockroachdb/cockroach/blob/cefa40d26fc4f0677b5a4e38c2d41d1ec23461e4/storage/sequence_cache.go#L110)):*
what's this local for?

---

*[storage/sequence_cache.go, line 114 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4526#-KB5i6a5RdplzH9ZoqNg:-KB5i6a5RdplzH9ZoqNh:1359251908) ([raw file](https://github.com/cockroachdb/cockroach/blob/cefa40d26fc4f0677b5a4e38c2d41d1ec23461e4/storage/sequence_cache.go#L114)):*
this is probably not needed, right? i'd expect the caller to not use the values without checking the error. this would then move to the single `false, nil` case. i don't feel strongly, though.

---

*[storage/sequence_cache.go, line 140 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4526#-KB5hsWvW1u0dL9ljw9q:-KB5hsWvW1u0dL9ljw9r:-555497277) ([raw file](https://github.com/cockroachdb/cockroach/blob/cefa40d26fc4f0677b5a4e38c2d41d1ec23461e4/storage/sequence_cache.go#L140)):*
probably meant to remove these

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4526)*
<!-- Sent from Reviewable.io -->

petermattis commented:
LGTM, though I'm not terribly familiar with this code. Storing the transaction isolation level in the intent seems fine. That's only an extra 1-2 bytes or so per intent when encoded.

---

Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4526#-:-KB9Qmkav4gOotB2ecq5:-1434085622)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Ok, I'll add this to the proto instead.

On Mon, Feb 22, 2016 at 2:25 PM Peter Mattis <notifications@github.com>
wrote:

> LGTM, though I'm not terribly familiar with this code. Storing the
> transaction isolation level in the intent seems fine. That's only an extra
> 1-2 bytes or so per intent when encoded.
> ------------------------------
>
> Review status: all files reviewed at latest revision, 5 unresolved
> discussions, some commit checks failed.
> ------------------------------
>
> *Comments from the review on Reviewable.io
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/4526#-:-KB9Qmkav4gOotB2ecq5:-1434085622>*
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/4526#issuecomment-187330454>
> .
>

tschottdorf commented:
Updated, PTAL.

tamird commented:
Reviewed 11 of 12 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[storage/replica.go, line 1457 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4526#-KBPeZv4zO8vGksDlBRJ:-KBPeZv4zO8vGksDlBRK:-1409346083) ([raw file](https://github.com/cockroachdb/cockroach/blob/5528b96996a33fe37ec1e23a7da131975527adcc/storage/replica.go#L1457)):*
s/restart/restarted/?

---

*[storage/replica_command.go, line 1049 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4526#-KBPfXQ4OGaRVrU3l2Jq:-KBPfXQ4OGaRVrU3l2Jr:2041339693) ([raw file](https://github.com/cockroachdb/cockroach/blob/5528b96996a33fe37ec1e23a7da131975527adcc/storage/replica_command.go#L1049)):*
I think you forgot to update this comment.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4526)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
LGTM

---

Review status: 10 of 12 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[storage/replica_command.go, line 1066 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4526#-KBPo6GLyntG9JIbSlIm:-KBPo6GLyntG9JIbSlIn:167348644) ([raw file](https://github.com/cockroachdb/cockroach/blob/5528b96996a33fe37ec1e23a7da131975527adcc/storage/replica_command.go#L1066)):*
`poison != 0` so I don't have to go back and see that it's unsigned or wonder what a negative value might mean.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4526#-:-KBPpBtGtTYOR4KN8_9W:676834233)*
<!-- Sent from Reviewable.io -->

tbg committed:
Reintroduce poisoning on transaction restarts

When pushing the intent of a non-SNAPSHOT transaction, we now poison the
response cache to force the transaction to retry should it revisit the Range.

This necessitated moving the `Isolation` field onto `TxnMeta` as per
discussions on the previous iteration of the corresponding PR #4526.
